staging/src/k8s.io/client-go/tools/watch/retrywatcher.go

```diff
@@ -0,0 +1,178 @@
+type RetryWatcherError struct {
+	s string
+}
+
+// Error implements error interface.
+func (e RetryWatcherError) Error() string {
+	return e.s
+}
+
+// GetObjectKind implements runtime.Object interface.
+func (e RetryWatcherError) GetObjectKind() schema.ObjectKind { return schema.EmptyObjectKind }
+
+// DeepCopyObject implements runtime.Object interface.
+func (e RetryWatcherError) DeepCopyObject() runtime.Object { return e }
+
+var _ error = RetryWatcherError{}
+var _ runtime.Object = RetryWatcherError{}
+
+// resourceVersionGetter is an interface used to get resource version from events.
+// We can't reuse an interface from meta otherwise it would be a cyclic dependency and we need just this one method
+type resourceVersionGetter interface {
+	GetResourceVersion() string
+}
+
+// RetryWatcher will make sure that in case the underlying watcher is closed (e.g. due to API timeout or etcd timeout)
+// it will get restarted from the last point without the consumer even knowing about it.
+// RetryWatcher does that by inspecting events and keeping track of resourceVersion.
+// Especially useful when using watch.Until where premature termination is causing troubles and flakes.
+// Please note that this is not resilient to ETCD cache not having the resource version anymore - you would need to
+// use Informers for that.
+type RetryWatcher struct {
+	lastResourceVersion string
+	watcherFunc         WatcherFunc
+	resultChan          chan watch.Event
+	stopChan            chan struct{}
+}
+
+// NewRetryWatcher creates a new RetryWatcher.
+// It will make sure that watcher gets restarted in case of recoverable errors.
+// The initialResourceVersion will be given to watchFunc when first called.
+func NewRetryWatcher(initialResourceVersion string, watcherFunc WatcherFunc) *RetryWatcher {
+	rw := &RetryWatcher{
+		lastResourceVersion: initialResourceVersion,
+		watcherFunc:         watcherFunc,
+		stopChan:            make(chan struct{}),
+		resultChan:          make(chan watch.Event, 0),
+	}
+	go rw.receive()
+	return rw
+}
+
+// doReceive returns true when it is done, false otherwise
+func (rw *RetryWatcher) doReceive() bool {
+	watcher := rw.watcherFunc(rw.lastResourceVersion)
+	if watcher == nil {
+		rw.resultChan <- watch.Event{
+			Type:   watch.Error,
+			Object: RetryWatcherError{fmt.Sprintf("RetryWatcher failed: watcherFunc returned nil watcher!")},
+		}
+		// Stop the watcher
+		return true
+	}
+	ch := watcher.ResultChan()
+	defer watcher.Stop()
+
+	for {
+		select {
+		case <-rw.stopChan:
+			glog.V(5).Info("Stopping RetryWatcher.")
+			return true
+		case event, ok := <-ch:
+			if !ok {
+				glog.Warningf("RetryWatcher - getting event failed! Re-creating the watcher. Last RV: %s", rw.lastResourceVersion)
+				return false
+			}
+
+			// We need to inspect the event and get ResourceVersion out of it
+			switch event.Type {
+			case watch.Added, watch.Modified, watch.Deleted:
+				metaObject, ok := event.Object.(resourceVersionGetter)
```

Comment:
In general yes.
But, can we ensure that those won't be coming (via some convoluted path) from informer?
It may not be possible now, but we may add some strange path in the future...

Reply:
I started implementing this as it seemed pretty easy, but then I've found out that the interface of event requires `runtime.Object` and `cache.DeletedFinalStateUnknown` doesn't implement this interface. At that point future proofing get kind of weird with the extra `interface{}` cast.
```golang
tombstone, ok := event.Object.(interface{}).(cache.DeletedFinalStateUnknown)
```

I don't think that's a good fit.

Reply:
agree we don't need to check that here when dealing with something that conforms to the watcher interface, but we do in NewInformerWatcher

Reply:
I've added the check there, thx